### ISSUE TITLE
Categorical columns with same name get read wrong

### ISSUE THREAD

**Issue by ivirshup** [theislab/anndata#147]
If you have categorical columns with the same name in obs and var, your h5ad overwrites the obs categories with the var categories. @Koncopd this seems relevant to any changes you'll be making to how data frames are written.

The case that led me to this:

```python
import anndata
import pandas as pd
import numpy as np

orig = anndata.AnnData(
    X=np.ones((2, 1)),
    obs=pd.DataFrame({"cat": pd.Categorical(["a", "b"])}),
    var=pd.DataFrame({"cat": pd.Categorical(["a"])})
)
orig.write("tmp.h5ad")
curr = anndata.read("tmp.h5ad")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-0c23bb631d15> in <module>
      6 )
      7 orig.write("tmp.h5ad")
----> 8 curr = anndata.read("tmp.h5ad")

~/github/anndata/anndata/readwrite/read.py in read_h5ad(filename, backed, chunk_size)
    456     else:
    457         # load everything into memory
--> 458         constructor_args = _read_args_from_h5ad(filename=filename, chunk_size=chunk_size)
    459         X = constructor_args[0]
    460         dtype = None

~/github/anndata/anndata/readwrite/read.py in _read_args_from_h5ad(adata, filename, mode, chunk_size)
    511     if not backed:
    512         f.close()
--> 513     return AnnData._args_from_dict(d)
    514 
    515 

~/github/anndata/anndata/base.py in _args_from_dict(ddata)
   2082                         d_true_keys[ann][k_stripped] = pd.Categorical.from_codes(
   2083                             codes=d_true_keys[ann][k_stripped].values,
-> 2084                             categories=v,
   2085                         )
   2086                 k_to_delete.append(k)

/usr/local/lib/python3.7/site-packages/pandas/core/arrays/categorical.py in from_codes(cls, codes, categories, ordered, dtype)
    659         if len(codes) and (
    660                 codes.max() >= len(dtype.categories) or codes.min() < -1):
--> 661             raise ValueError("codes need to be between -1 and "
    662                              "len(categories)-1")
    663 

ValueError: codes need to be between -1 and len(categories)-1
```

But if we flip obs and var (edited with better example):

```python
orig = anndata.AnnData(
    X=np.ones((2, 2)),
    obs=pd.DataFrame({"cat": pd.Categorical(["a", "b"])}),
    var=pd.DataFrame({"cat": pd.Categorical(["c", "d"])}),
)
orig.write("tmp.h5ad")
curr = anndata.read("tmp.h5ad")
curr.obs["cat"].cat.categories
# Index(['c', 'd'], dtype='object')
```

**Comment by falexwolf** [comment]
Yes, I know, that's poor design. That's why the suggestion of storing `DataFrame` in one place in the hdf5 (e.g., one HDF5 group as for `anndata.h5py.SparseDataset`) would make much more sense.

Are you working on this @Koncopd?

**Comment by Koncopd** [comment]
@falexwolf yes, i do.

**Comment by ivirshup** [comment]
Fixed by #167
